SCC-2206 Add Holdings Parser

mwbenowitz opened the pull request:
This creates a set of new classes and functions for handling Holding records coming from the new Holding pipeline. These represent serial holdings that are to be represented in the SCC front-end.

These records function very similarly to Item records, in that they are related to Bibs in a (generally) one-to-many relationship. The records themselves are fairly simple and the a new mapping is being created in NYPL-core to represent them.

In addition to adding the new Holding updater this adds some new logic to the index to properly route all new records.

No new tests have been created yet, these will be added soon, but wanted to validate this general approach to extending the store-poster before delving into those issues. I've commented in the PR with various questions I have.

One big question is regarding the `Updater` code. Bib, Item and Holding records all have their own updaters with a very basic base class with a single method. In writing the Holding updater I noticed that these three classes could probably share much of their code, but that would be a fairly major refactor and I'm not sure what the attitude towards that is since this is fairly stable already.

mwbenowitz commented:
Question for folks here: I can't figure out why one test passed and one failed. As far as I can tell in one it seems like the `chalk` dependency is trying to use non-es6 syntax? Any ideas?

mwbenowitz commented on index.js:
```diff
@@ -40,27 +47,28 @@ function getSchema (schemaName) {
 
     // Invoke appropriate updater:
-    var updater = bibOrItem === 'Bib' ? (new BibsUpdater()) : (new ItemsUpdater())
+    var updater = new updaters[bibItemOrHolding]()
```
I'm not 100% sure this syntax will work, and the process of writing unit tests for the handler is proving to be tough. Yay or nay on if it will work?

danamansana replied:
It should work

mwbenowitz commented on lib/holdings-updater.js:
```diff
@@ -0,0 +1,295 @@
+
+const CACHE_TO_DISK = false
+
+class HoldingsUpdater extends UpdaterBase {
```
This is the class that shares a lot of functionality with `BibsUpdater` and `ItemsUpdater`. I'm wondering if it makes sense to try to factor out some of those redundancies, or if that brings up too much risk of regressions. As there are three classes now the potential advantages are greater.

mwbenowitz replied:
That sounds like a good approach, definitely want to be careful with the code base because it seems very likely that decisions were made for reasons that I don't know or understand.

mwbenowitz commented on lib/holdings-updater.js:
```diff
@@ -0,0 +1,295 @@
+  }
+
+  saveToDisc (holding) {
+    const path = `./cache/holdings/${holding.id}.json`
```
There is no `nyplSource` in the holdings record, so I've eliminated that wherever it appears. I don't think that will have any side effects. (Here for items it is `./cache/items/${nyplSource/i${item.id}.json`

mwbenowitz commented on lib/nyplDataApi.js:
```diff
@@ -26,4 +26,8 @@ const bib = (source, id) => {
 
-module.exports = { bib, item }
+const holding = (id) => {
+  return client().get(`holdings?ids=${id}`)[0]
```
There are dedicated `GET` endpoints for `Bib` and `Item` records that retrieve a single record. When setting up the `HoldingService` it was decided that we didn't need an endpoint that replicated that functionality. 

Technically we don't, since this should work fine, but it would be a little cleaner if we did. Not sure if it makes sense to add it just for this single use case, but it would make this cleaner and align it with the bib and item methods

mwbenowitz replied:
I agree, in general I think there is value in aligning all of these services, it shouldn't be hard, and it is easy to imagine other services having this need in the future.

@danamansana Could you create a ticket to add this endpoint to the `HoldingService`? We should probably try to get to it this week or next, but not immediately necessary.

mwbenowitz commented on lib/serializers/holding.js:
```diff
@@ -0,0 +1,163 @@
+  try {
+    const fieldMapper = buildFieldMapper('holding', null)
+
+    const builder = Statement.builder(object.id, Creator.CORE_SERIALIZER.id, { record_id: object.id })
```
Bib and Item ids are prefixed (the `b` and `i` standard prefixes we all know and love), but generally we don't have a prefix for the holdings records. I'm not sure if it would be a problem if we don't? Otherwise, we would need to add `h` as a prefix to all of the identifiers used here.

I have no real preference other than a desire to keep things simple. I feel like we probably should but want to get input from others first.

nonword replied:
You're proposing not prefixing holding ids? Everything in the DiscoveryStore exists in the same id-space. All records are statements about "resources". There's thus a technical limitation that whatever id you use be unique in the universe of bib and item ids. Because the latter have used prefixes, you could technically not use prefixes for holdings, but we'd only get to do that for one resource type. Short term, I'd use 'h' as a prefix (or 'holding:'). Medium term, maybe we need a dedicated "resource_statement" table for each type (i.e. bib_statement, item_statement, holding_statement). Long term, the DiscoveryStore and greater data pipeline should be rethought from a high level.

mwbenowitz replied:
ok yeah, I had a feeling that would be the answer after thinking about this more and definitely agree. Was trying to get away something that would have been too simple and almost certainly would have caused issues

nonword reviewed (commented):
Looking good! Some comments

mwbenowitz commented on lib/holdings-updater.js:
```diff
@@ -0,0 +1,295 @@
+  }
+}
+
+HoldingsUpdater.threaded = (opts) => {
```
yeah, I honestly thought this was used for things like loading test records, etc. so its good to know that it hasn't been maintained. I'll definitely strip it out because we can much more easily push records into the `Holding` stream if/when we need to bulk load holding records.

mwbenowitz commented:
@nonword @danamansana This should be ready for real review now. I've decided not to do any of the refactoring work at this point because it is time that would probably be better spent on the API side of things and fixing lingering issues in the Holding pipeline.

I simplified the parser process a bit and think I've updated everything to match Paul's earlier comments, so any thoughts would be greatly appreciated! And if there are any useful tests that this is missing let me know.

nonword commented on test/holding-marc-test.js:
```diff
@@ -0,0 +1,131 @@
+    })
+    return object
+  }
+ */
```
Looks like a useful utility function for testing, but is it used?

mwbenowitz replied:
oh yeah, its not, I borrowed it from one of the other test classses because it seemed like I might need it, but I can't think of a test case where I would. I'll just remove it.

nonword reviewed (approved):
Looks great
